File: settlement/room/food/orchard/Gui.java
package settlement.room.food.orchard;

import game.GAME;
import game.time.TIME;
import init.sprite.SPRITES;
import settlement.room.industry.module.IndustryResource;
import settlement.room.industry.module.IndustryUtil;
import settlement.room.main.RoomInstance;
import snake2d.util.datatypes.COORDINATE;
import snake2d.util.datatypes.DIR;
import snake2d.util.gui.GUI_BOX;
import snake2d.util.gui.GuiSection;
import snake2d.util.gui.clickable.CLICKABLE;
import snake2d.util.gui.renderable.RENDEROBJ;
import snake2d.util.sets.LISTE;
import util.data.GETTER;
import util.gui.misc.GBox;
import util.gui.misc.GButt;
import util.gui.misc.GGrid;
import util.gui.misc.GHeader;
import util.gui.misc.GStat;
import util.gui.misc.GText;
import util.gui.table.GStaples;
import util.gui.table.GTableSorter.GTFilter;
import util.gui.table.GTableSorter.GTSort;
import util.info.GFORMAT;
import util.text.D;
import util.text.Dic;
import util.text.DicTime;
import view.sett.ui.room.UIRoomBulkApplier;
import view.sett.ui.room.UIRoomModule.UIRoomModuleImp;

class Gui extends UIRoomModuleImp<Instance, ROOM_ORCHARD> {

	static CharSequence ¤¤Trees = "¤Trees";
	static CharSequence ¤¤TreesD = "¤Amount of fully grown trees. Only when trees are fully grown will they start producing. Neglected trees will die.";
	private static CharSequence ¤¤TreeNext = "¤Next tree will be grown in:";
	
	
	private static CharSequence ¤¤estimated = "¤Estimated Harvest (year)";
	private static CharSequence ¤¤daysToHarvest = "¤Days until harvest";
	private static CharSequence ¤¤baseValue = "¤Capacity";
	
	private static CharSequence ¤¤HarvestYear = "¤This Year";
	private static CharSequence ¤¤HarvestPrev = "¤Last Year";

	private static CharSequence ¤¤skill = "¤Worker Skill";
	private static CharSequence ¤¤skillD = "¤The average bonus accumulated during the year. This determines the output of the harvest and the growth rate of the trees.";
	private static CharSequence ¤¤skillCurrent = "¤Bonus (current)";
	private static CharSequence ¤¤skillCurrentD = "¤The bonus that is currently being added to the Orchard.";
	
	private static CharSequence ¤¤chop = "¤Chop";
	private static CharSequence ¤¤chopD = "¤Reset all progress by chopping down the trees and instantly get {0} {1}.";
	
	Gui(ROOM_ORCHARD s) {
		super(s);
		D.t(this);
	}
	
	final Cache cache = new Cache();
	
	@Override
	public void hover(GBox box, Instance i) {
		super.hover(box, i);
		box.NL();
		if (!i.blueprintI().constructor.isIndoors) {
			box.text(blueprint.constructor.fertility.name());
			box.add(GFORMAT.perc(box.text(), blueprint.constructor.fertility.get(i)));
			
			box.NL();
		}
		box.text(¤¤estimated);
		box.add(GFORMAT.i(box.text(), (int)cache.output(i)));
		
		box.space();
	}
	
	@Override
	protected void appendMain(GGrid icons, GGrid text, GuiSection sExtra) {
		
		
		
		IndustryResource res = blueprint.industries().get(0).outs().get(0);
		
		text.add(new GHeader(Dic.¤¤Production));
		
		GuiSection s = new GuiSection() {
			
			@Override
			public void hoverInfoGet(GUI_BOX text) {
				text.title(¤¤estimated);
				
			};
			
		};
		
		s.add(res.resource.icon(), 0, 0);
		
		s.addRightC(4, new GStat() {
			
			@Override
			public void update(GText text) {
				int am = 0;
				for (int i = 0; i < blueprint.instancesSize(); i++) {
					Instance ins = blueprint.getInstance(i);
					am += cache.output(ins);
				}
				GFORMAT.i(text, am);
			}
		});
		
		text.add(s);
		
		GStaples st = new GStaples(res.history().historyRecords()) {
			
			@Override
			protected void hover(GBox box, int stapleI) {
				int i = res.history().historyRecords()-1-stapleI;
				int am = res.history().get(i);
				GText t = box.text();
				DicTime.setDaysAgo(t, i);
				box.add(t);
				box.NL(2);
				box.add(GFORMAT.i(box.text(), am));
			}
			
			@Override
			protected double getValue(int stapleI) {
				return res.history().get(res.history().historyRecords()-1-stapleI);
			}
		};
		

		st.body().setWidth(180).setHeight(64);
		text.add(st);
		
	}

	@Override
	protected void appendPanel(GuiSection section, GGrid grid, GETTER<Instance> getter, int x1, int y1) {
		
		GuiSection s = new GuiSection();
		
		s.add(prod(getter));
		
		RENDEROBJ o = new GStat() {
			@Override
			public void update(GText text) {
				int i = cache.daysTillNextTree(getter.get());
				if (i < Integer.MAX_VALUE) {
					int r = blueprint.time.ripeDay - (TIME.days().bitsSinceStart() + i) % (blueprint.time.ripeDay+1);
					DicTime.setDays(text, i + r);
				}
				else
					DicTime.setDays(text, blueprint.time.daysTillHarvest());
			}
		}.hh(SPRITES.icons().s.clock).hoverInfoSet(¤¤daysToHarvest);
		s.addRightC(32, o);
		s.body().incrW(64);
		
		s.addRelBody(4, DIR.N, new GHeader(Dic.¤¤Production));
		section.addRelBody(8, DIR.S, s);
		
		
		s = new GuiSection();
		
		int tab = 180;
		
		
		s.addDown(2, new GStat() {
			
			@Override
			public void update(GText text) {
				Instance ins = getter.get();
				GFORMAT.iofkInv(text, cache.trees(ins), cache.treesTotal(ins));
				
				if (cache.daysTillNextTree(ins) < Integer.MAX_VALUE) {
					text.s().add('(').s();
					DicTime.setDays(text, cache.daysTillNextTree(ins));
					text.s().add(')');
				}
				
			}
			
			@Override
			public void hoverInfoGet(GBox b) {
				b.title(¤¤Trees);
				b.text(¤¤TreesD);
				b.NL(8);
				
				if (cache.daysTillNextTree(getter.get()) < Integer.MAX_VALUE) {
					b.textL(¤¤TreeNext);
					GText t = b.text();
					DicTime.setDays(t, cache.daysTillNextTree(getter.get()));
					b.add(t);
				}
			};
			
		}.hh(¤¤Trees, tab).increaseWidth(100));
		
		s.addDown(2, new GStat() {
			
			@Override
			public void update(GText text) {
				GFORMAT.f1(text, getter.get().skill());
			}
			
			@Override
			public void hoverInfoGet(GBox b) {
				b.title(¤¤skill);
				b.text(¤¤skillD);
				b.NL(8);
				b.textLL(Dic.¤¤Value);
				b.tab(6);
				b.add(GFORMAT.f1(b.text(), getter.get().skill()));
				
				b.sep();
				b.textLL(¤¤skillCurrent);
				b.NL();
				b.text(¤¤skillCurrentD);
				b.NL();
				IndustryUtil.hoverBoosts(b, 1.0, getter.get().industry(), getter.get().industry().bonus(), getter.get(), 1);
				
				
			};
			
		}.hh(¤¤skill, tab).increaseWidth(100));
		
		section.add(s, section.body().x1(), section.body().y2()+16);
		
		
		CLICKABLE c = new GButt.ButtPanel(¤¤chop) {
			
			@Override
			protected void renAction() {
				activeSet(cache.wood > 0);
			};
			
			@Override
			protected void clickA() {
				
				Instance ins = getter.get();
				
				for (COORDINATE c : ins.body()) {
					if (!ins.is(c))
						continue;
					OTile t = blueprint.tile.getM(c.x(), c.y());
					if (t != null) {
						t.chop();
					}
				}
			}
			
			@Override
			public void hoverInfoGet(GUI_BOX text) {
				GBox b = (GBox) text;
				GText t = b.text();
				t.add(¤¤chopD);
				t.insert(0, cache.wood);
				t.insert(1, blueprint.auxRes.resource().names);
				b.add(t);
			};
			
		}.pad(8, 4);
		
		section.addRelBody(16, DIR.S, c);
		
	}
	
	private RENDEROBJ prod(GETTER<Instance> getter) {
		GuiSection s = new GuiSection() {
			
			@Override
			public void hoverInfoGet(GUI_BOX text) {
				GBox b = (GBox) text;
				Instance ins = getter.get();

				b.textL(¤¤baseValue);
				b.tab(6);
				b.add(GFORMAT.f(b.text(), ins.base));
				b.NL();
				
				b.textL(Dic.¤¤ProductionRate);
				b.tab(6);
				b.add(GFORMAT.f(b.text(), blueprint.productionData.outs().get(0).rate));
				b.NL();
				
				b.textL(¤¤skill);
				b.tab(6);
				b.add(GFORMAT.f(b.text(), ins.skill()));
				b.NL();

				b.textLL(Dic.¤¤Total);
				b.tab(6);
				b.add(GFORMAT.f1(b.text(), ins.base*blueprint.productionData.outs().get(0).rate*ins.skill()));
				b.NL();
				
				b.textL(¤¤Trees);
				b.tab(6);
				b.add(GFORMAT.f(b.text(), (double)ins.trees/ins.treesTotal));
				b.NL();
				
				b.textLL(¤¤estimated);
				b.tab(6);
				b.add(GFORMAT.f1(b.text(), cache.output(ins)));
				b.NL();
				
				b.NL(4);
				
	
				
				b.NL(16);
				b.textL(¤¤HarvestYear);
				b.tab(6);
				b.add(GFORMAT.i(b.text(), (int)ins.blueprintI().indus.get(0).outs().get(0).year.get(ins)));
				b.NL();
				
				b.NL(2);
				b.textL(¤¤HarvestPrev);
				b.tab(6);
				b.add(GFORMAT.i(b.text(), (int)ins.blueprintI().indus.get(0).outs().get(0).yearPrev.get(ins)));
				b.NL();
				
			}
			
		};
		
		s.add(blueprint.productionData.outs().get(0).resource.icon(), 0, 0);
		GStat stat = new GStat() {
			
			@Override
			public void update(GText text) {
				double am = cache.output(getter.get())/TIME.years().bitConversion(TIME.days());
				GFORMAT.f0(text, am);
			}
		};
		
		s.addRightC(6, stat);
		
		s.body().incrW(64);
		return s;
	}
	
	@Override
	protected void appendTableFilters(LISTE<GTFilter<RoomInstance>> filters, LISTE<GTSort<RoomInstance>> sorts,
			LISTE<UIRoomBulkApplier> appliers) {
	}
	
	private class Cache {
		
		private int upI = -1;
		private int treesTotal;
		private int trees;
		private int daysTillNextTree;
		private int wood;
//		private double fertilityNext;
		private double output;
		
		private Instance ins;
		
		public int treesTotal(Instance ins) {
			up(ins);
			return treesTotal;
		}
		
		public int trees(Instance ins) {
			up(ins);
			return trees;
		}
		
		public int daysTillNextTree(Instance ins) {
			up(ins);
			return daysTillNextTree;
		}
		
		public double output(Instance ins) {
			up(ins);
			return output;
		}
		
		private void up(Instance ins) {

			
			
			if (upI == GAME.updateI() && this.ins == ins)
				return;
			this.ins = ins;
			upI = GAME.updateI();
			
			wood = 0;
			//fertilityNext = 0;
			treesTotal = 0;
			trees = 0;
			daysTillNextTree = Integer.MAX_VALUE;
			
			for (COORDINATE c : ins.body()) {
				if (!ins.is(c))
					continue;
				
				OTile t = blueprint.tile.getM(c.x(), c.y());
				if (t != null) {
					treesTotal ++;
					if (t.state() == t.IBIG)
						trees ++;
					else {
						int d = t.state().daysTillGrown();
						if (d < daysTillNextTree) {
							
							daysTillNextTree = d;
						}
					}
					if (t.state() == t.ISMALL )
						wood+= blueprint.auxRes.amount()/2;
					else if (t.state() == t.IBIG || t.state() == t.IDEAD)
						wood+= blueprint.auxRes.amount();
				}

			}
			
			output = blueprint.time.days*ins.skill()*ins.base*ins.industry().outs().get(0).rate*ins.trees/ins.treesTotal;
		}
		
	}


}

File: LegacySource/settlement/room/infra/station/Gui.cs
using System;
using System.Collections.Generic;
using System.Linq;
using game;
using init.resources;
using init.settings;
using init.sprite.UI;
using settlement.entity.humanoid;
using settlement.main;
using settlement.room.infra.transport;
using settlement.room.main;
using snake2d;
using snake2d.util.color;
using snake2d.util.datatypes;
using snake2d.util.gui;
using snake2d.util.gui.GuiSection;
using snake2d.util.gui.Hoverable;
using snake2d.util.gui.renderable;
using snake2d.util.rnd;
using snake2d.util.sets;
using util.colors;
using util.data;
using util.gui.misc;
using util.gui.slider;
using util.gui.table;
using util.info;
using util.text;
using view.main;
using view.sett.ui.room;

class Gui : UIRoomModuleImp<StationInstance, ROOM_STATION>
{
    private static readonly CharSequence ¤Accepting = "Accepting";
    private static readonly CharSequence ¤Capacity = "Capacity";
    private static readonly CharSequence ¤Crates = "Crates";
    private static readonly CharSequence ¤Incoming = "Incoming";
    private static readonly CharSequence ¤Reserved = "Reserved";
    private static readonly CharSequence ¤Resources = "Resources";
    private static readonly CharSequence ¤Stored = "Stored";

    private int bonusAccepting;
    private int bonusIncoming;

    public Gui() : base(new GuiSection())
    {
        var g = new GuiSection();
        g.add(new GHeader(¤Resources));
        g.add(new GScrollRows(GenerateResourceLines(), 350));
        g.add(new GStat(() => $"Total Crates: {g.get().crates.size()}"));
        g.pad(8, 8);

        body().add(g);
    }

    private List<RENDEROBJ> GenerateResourceLines()
    {
        var lines = new List<RENDEROBJ>();

        foreach (var res in RESOURCES.ALL())
        {
            lines.Add(new ResLine(res, g => g.get().tally(res)));
        }

        return lines;
    }

    private class ResLine : HoverableAbs
    {
        private readonly RESOURCE res;
        private readonly Func<StationInstance, StationTally> tallyGetter;

        public ResLine(RESOURCE res, Func<StationInstance, StationTally> tallyGetter) : base(300, 32)
        {
            this.res = res;
            this.tallyGetter = tallyGetter;
        }

        protected override void render(SPRITE_RENDERER r, float ds, bool isHovered)
        {
            res.icon().renderCY(r, body.x1(), body.cY());
            var ins = g.get();
            var t = tallyGetter(ins);

            int x2 = body.x2() - 48;
            GMeter.render(r, GMeter.C_REDGREEN, (double)t.stored() / t.space(), (double)(t.stored() + ins.incoming(res)) / t.space(), body.x1() + 30, x2, body.y1() + 4, body.y2() - 4);

            Str.TMP.clear().add(t.stored());
            int w = UI.FONT().S.width(Str.TMP);
            OPACITY.O50.bind();
            COLOR.BLACK.render(r, x2 - w - 8, x2, body.y1() + 5, body.y2() - 5);
            OPACITY.unbind();

            UI.FONT().S.renderCY(r, x2 - w - 4, body.cY(), Str.TMP);

            if (ins.accepting(res))
            {
                GCOLOR.T().IGOOD.bind();
            }
            else
            {
                GCOLOR.T().IBAD.bind();
            }
            UI.icons().s.storage.renderCY(r, x2 + 8, body.cY());

            if (SETT.ROOMS().TRANSPORT.hasActive(res))
            {
                GCOLOR.T().IGOOD.bind();
            }
            else
            {
                GCOLOR.T().IBAD.bind();
            }
            UI.icons().s.chevron(DIR.E).renderCY(r, x2 + 24, body.cY());
            COLOR.unbind();
        }

        public override void hoverInfoGet(GUI_BOX text)
        {
            GBox b = (GBox)text;
            b.title(res.name);

            b.textLL(¤Stored);
            b.tab(6);
            b.add(GFORMAT.i(b.text(), tallyGetter(g.get()).stored()));
            b.NL();

            b.textLL(¤Capacity);
            b.tab(6);
            b.add(GFORMAT.i(b.text(), tallyGetter(g.get()).space()));
            b.NL();

            b.textLL(¤Reserved);
            b.tab(6);
            b.add(GFORMAT.i(b.text(), tallyGetter(g.get()).reserved()));
            b.NL();

            b.textLL(¤Incoming);
            b.tab(6);
            b.add(GFORMAT.i(b.text(), g.get().incoming(res)));
            b.NL();

            if (g.get().accepting(res))
            {
                b.add(b.text().normalify2().add(¤Accepting));
            }
            else
            {
                b.error(¤AcceptingNot);
            }
            if (S.get().developer)
                b.add(b.text().add(g.get().blueprintI().tally(res).accepting()));
            b.NL();

            if (SETT.ROOMS().TRANSPORT.hasActive(res))
            {
                b.add(b.text().normalify2().add(¤Has));
            }
            else
            {
                b.error(¤HasNot);
            }
            b.NL();

            base.hoverInfoGet(text);
        }
    }
}